js.go

```diff
@@ -1442,10 +1463,41 @@ func (js *js) subscribe(subj, queue string, cb MsgHandler, ch chan *Msg, isSync,
 	if hasHeartbeats {
 		sub.scheduleHeartbeatCheck()
 	}
+	// For ChanSubscriptions, if we know that there is flow control, we will
+	// start a go routine that evaluates the number of delivered messages
+	// and process flow control.
+	if sub.Type() == ChanSubscription && hasFC {
+		go sub.chanSubcheckForFlowControlResponse()
+	}
 
 	return sub, nil
 }
 
+// This long-lived routine is used per ChanSubscription to check
+// on the number of delivered messages and check for flow control response.
+func (sub *Subscription) chanSubcheckForFlowControlResponse() {
+	sub.mu.Lock()
+	nc := sub.conn
+	sub.mu.Unlock()
+	if nc == nil {
+		return
+	}
+	t := time.NewTicker(chanSubFCCheckInterval)
+	for range t.C {
+		sub.mu.Lock()
+		if sub.closed {
+			sub.mu.Unlock()
+			t.Stop()
+			return
+		}
+		fcReply := sub.checkForFlowControlResponse()
+		sub.mu.Unlock()
+		if fcReply != _EMPTY_ {
```

Comment:
If we don't check for errors just send blindly? No need to check fcReply?

Reply:
Well, sure, Conn.Publish() checks for empty subject and returns an error, but I thought that it makes more readable that we know that fcReply will be non empty only when we need to reply back to FC.